admin/server/postgres.go

```diff
@@ -0,0 +1,204 @@
+			return writer.Complete(strings.Trim(upperQuery, ";"))
+		}, wire.WithColumns(nil))), nil
+	}
+
+	clientParams := wire.ClientParameters(ctx)
+	// database is org.password
+	tokens := strings.Split(clientParams[wire.ParamDatabase], ".")
+	if len(tokens) != 2 {
+		return nil, fmt.Errorf("invalid org or project")
+	}
+	org := tokens[0]
+	project := tokens[1]
+
+	// Find the production deployment for the project we're proxying to
+	proj, err := s.admin.DB.FindProjectByName(ctx, org, project)
+	if err != nil {
+		return nil, fmt.Errorf("invalid org or project")
+	}
+
+	if proj.ProdDeploymentID == nil {
+		return nil, fmt.Errorf("no prod deployment for project")
+	}
+	depl, err := s.admin.DB.FindDeployment(ctx, *proj.ProdDeploymentID)
+	if err != nil {
+		return nil, fmt.Errorf("no prod deployment for project")
+	}
+
+	var jwt string
+	claims := auth.GetClaims(ctx)
+	switch claims.OwnerType() {
+	case auth.OwnerTypeAnon:
+		// If the client is not authenticated with the admin service, we just proxy the contents of the password to the runtime (if any).
+		password := ctx.Value(auth.PostgresPassword{}).(string)
+		if len(password) >= 6 && strings.EqualFold(password[0:6], "bearer") {
+			jwt = strings.TrimSpace(password[6:])
+		}
+	case auth.OwnerTypeUser, auth.OwnerTypeService:
+		// If the client is authenticated with the admin service, we issue a new ephemeral runtime JWT.
+		// The JWT should have the same permissions/configuration as one they would get by calling AdminService.GetProject.
+		permissions := claims.ProjectPermissions(ctx, proj.OrganizationID, depl.ProjectID)
+		if !permissions.ReadProd {
+			return nil, fmt.Errorf("does not have permission to access the production deployment")
+		}
+
+		var attr map[string]any
+		if claims.OwnerType() == auth.OwnerTypeUser {
+			attr, err = s.jwtAttributesForUser(ctx, claims.OwnerID(), proj.OrganizationID, permissions)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		jwt, err = s.issuer.NewToken(runtimeauth.TokenOptions{
+			AudienceURL: depl.RuntimeAudience,
+			Subject:     claims.OwnerID(),
+			TTL:         runtimeAccessTokenDefaultTTL,
+			InstancePermissions: map[string][]runtimeauth.Permission{
+				depl.RuntimeInstanceID: {
+					// TODO: Remove ReadProfiling and ReadRepo (may require frontend changes)
+					runtimeauth.ReadObjects,
+					runtimeauth.ReadMetrics,
+					runtimeauth.ReadProfiling,
+					runtimeauth.ReadRepo,
+					runtimeauth.ReadAPI,
+				},
+			},
+			Attributes: attr,
+		})
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("runtime proxy not available for owner type %q", claims.OwnerType())
+	}
+
+	// Track usage of the deployment
+	s.admin.Used.Deployment(depl.ID)
+
+	hostURL, err := url.Parse(depl.RuntimeHost)
+	if err != nil {
+		return nil, err
+	}
+	hostURL.Scheme = "postgres"
+	hostURL.Host = hostURL.Hostname() + ":" + strconv.FormatInt(int64(15432), 10)
```

Comment:
We are already using similar hacks to determine grpc port. How about we change the port to `15432` if host contains `localhost`.

Reply:
I think those hacks are CLI/frontend side, so don't have access to `.env` vars. Reading `RILL_RUNTIME_PSQL_PORT` is a minor improvement (since `.env` is read in this process), but yeah only doing it if host contains `localhost` is fine